### ISSUE TITLE
멘토링 질문

### ISSUE THREAD

**Issue by heehehe** [heehehe/pruning-decorrelation#6]
- **기존 resnet에서 추가된 groups, dilation의 역할**
   - [pytorch conv2d 설명](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html)
   - groups : https://wingnim.tistory.com/104
   - dilation : https://zzsza.github.io/data/2018/02/23/introduction-convolution/
- **mnn.py 안의 Masker를 통해 MaskConv2d에서 masking이 될 때, mask 값 계산 과정**
   - main.py에서 args.prune 살펴보기
    [main.py#L58-64](https://github.com/INCHEON-CHO/Dynamic_Model_Pruning_with_Feedback/blob/6bdb54c886b6ca4837696695b0d957ad9ab5c2fc/main.py#L58-L64)
    ```
    elif args.prune:    # for pruning
        pruner = pruning.__dict__[args.pruner]
        model, image_size = pruning.models.__dict__[args.arch](data=args.dataset, num_layers=args.layers,
                                                   width_mult=args.width_mult,
                                                   depth_mult=args.depth_mult,
                                                   model_mult=args.model_mult,
                                                   mnn=pruner.mnn)
    ```   
   https://github.com/heehehe/pruning-decorrelation/blob/a6049d65caf5e2050b5fa215b0b722cbc91096d4/pruning/dpf/mnn.py#L22

**Comment by heehehe** [comment]
> * **기존 resnet에서 추가된 groups, dilation의 역할**
- groups : https://wingnim.tistory.com/104
- dilation : https://zzsza.github.io/data/2018/02/23/introduction-convolution/

**Comment by heehehe** [comment]
- 직교성과 대조적으로 correlation을 줄이는 식에 대한 내용을 loss로 추가해야 함
- correlation matrix의 상삼각행렬 절댓값의 합에 대한 내용을 W^T W - I 대신에 추가
- 직교성에 대한 연구는 이미 진행되었고, 이와 다른 방법으로 처리하는 것이 decorrelation 방법

**Comment by heehehe** [comment]
#### 10/23 멘토링
- orthogonal : loss term에 W^TW-I 더하기 (이미 진행된 연구)
- decorrelation : covariance의 상삼각행렬 절댓값의 합으로 
                          (corr으로도 해보기 - 그런데 분모가 0이 될수도 있어서 gradient exploding 일어날수도..?)
- Loss_t (classification loss) 에 orthogonal과 correlation 비교 실험
- 여기에서는 FLOPs 처리보다는 sparsity로 지표 측정하는 것이 효과적
- top1 : 상위 1개 / top5 : 상위 5개 중에 정답 맞췄는지
- learning rate, epoch 등은 가장 잘 나오는 것으로 설정되어 있는 것 + prune rate만 조절해서 실험해보기

**Comment by heehehe** [comment]
- pruning 에 걸어지는 제약조건 무엇인지
```if (globals()['iterations']+1) % args.prune_freq==0 and (epoch+1) <= args.milestones[1]:```
https://github.com/INCHEON-CHO/Dynamic_Model_Pruning_with_Feedback/blob/6bdb54c886b6ca4837696695b0d957ad9ab5c2fc/main.py#L244
: prune_freq마다 가지치기 되도록 하고, epoch이 milestones[1] 값을 넘으면 더 이상 가지치기 되지 않도록
(코드가 실행되지 않던 원인은 importance_all 대신 importance로 코드가 잘못 작성되어 있어서였음)
